### DISCUSSION TITLE
Include languages other than english for Spellcheck

### DISCUSSION THREAD

**Discussion by diseno-franco** [vkbo/novelWriter#737] [Help]
* A clear and concise description of what the problem is: I've been trying to add a spanish dictionary to the spellchecker, can't seem to find how to do it or where to add it
* A clear and concise description of what you want to happen: I want the spellchecker to spellcheck my spanish
* A clear and concise description of any alternative solutions or features you've considered: mixed language spellcheck and grammarcheck would be cool
* If applicable, add screenshots to help explain your request.

**Comment by vkbo** [comment]
Spell checking is provided by the [pyenchant](https://pypi.org/project/pyenchant/) library. It should look for dictionaries installed on your system, but you didn't specify which operating system you use. On Linux it's as simple as installing new hunspell dictionaries from the package manager.

novelWriter just tells pyenchant what dictionary to pick, that's all. You can set a global one in Preferences, and even set a custom one for a specific project in project settings.

As for grammar checking and other language tools, see #515 and #712 where there's an ongoing discussion on such features. I'm considering adding LanguageTools for instance.

**Comment by vkbo** [comment]
Also, you may have to select "Spell Enchant" as the spell check provider in Preferences. If it's set to "Internal", only English is available by default.

**Comment by vkbo** [comment]
Did you manage to add Spanish? I see from another issue you posted that you're using Windows, as I suspected from the question. My first language is Norwegian, so I've tested installing additional dictionaries on Linux which is my main OS. But I've not tried on Windows. I checked that pyenchant only comes with English by default on Windows.

Enchant isn't really built for Windows, although pyenchant provides binaries for it. I'm not sure how it pulls in additional dictionaries. Maybe it is possible to automate the process somehow. I don't have access to Windows right now, but I'll look into if.

**Reply by diseno-franco** [reply]
Thank you so much, I haven't really looked into it.  (Might I ad you are really on top of everything here, it's very impressive). Also if I want to help with the translation of NW do I just copy the file in dev and start or what should I do beforehand. I'm also a UX/UI desingner, in case there is a place where I could collaborate with GUIs and the like.)

**Reply by vkbo** [reply]
Getting a Spanish translation would be great! There is a translation issue thread at #93 if you want to contribute.

The i18n implementation currently lives in the `testing` branch and is in beta. The instructions for how to add translations is here: https://github.com/vkbo/novelWriter/tree/testing/i18n

Essentially, you just add the Spanish entry to the `novelWriter.pro` file and run the `setup.py qtlupdate` command to generate the translation-ready file. The translation itself is easiest done with the Qt Linguist application that is part of the Qt Creator suite. The file is an XML file with the `.ts` file extension.

**Comment by Geobert** [comment]
Hi, I'm facing this issue as well :-/ 

I'm using Windows 10 in UK english, but I also have French dictionnary installed but it doesn't appear in the dropbox :-/
Is there any way to force pyenchant to find a language?

**Reply by vkbo** [reply]
Apparently you need to get the relevant hunspell dictionaries.
See: http://pyenchant.github.io/pyenchant/install.html#installing-a-dictionary

I'll look into finding a way to simplify this process on Windows.

**Reply by vkbo** [reply]
I was looking into this today actually, and added an issue on this as #982

I wrote some instructions in the issue for how to add dictionaries on Windows 10. It's fairly straightforward. 

You can downloaded the *.dic and *.aff files for the language you want from: https://cgit.freedesktop.org/libreoffice/dictionaries/tree/

Then you make a folder called `hunspell` inside `C:\Users\<UserName>\AppData\Local\enchant` and add them there. Then they should show up in Preferences. At least it worked on my Windows 10 virtual machine.

I'm planning to add a simple way inside of novelWriter to import downloaded dictionaries so you don't have to do this manually.

**Reply by Ellipsine** [reply]
Thank you, it works ! 
Even if the french corrector identifies simple french words like "head" "to put" as faults. It seems to have even less vocabulary than a toddler. What the f… I had the same problems with this dictionaries in other softwares, it is so tiring.  Do you know other sources for french ? I never could find something that works. 

@jyhelle I think you speak and write in french, don't you ? Did you find a dictionarie that would have more words than… that ?

**Reply by vkbo** [reply]
The standard spell checking is done with hunspell on Windows (also aspell on Linux), which is the same spell checking library and dictionaries used by Libre Office, Open Office, Free Office, Firefox, Thunderbird, and I don't know what else. novelWriter only integrates with it. No actual spell checking happens in novelWriter code any more.

Some of the dictionaries really are incomplete. There are some very notable omissions for Norwegian too, like the word «noen» which means «somebody». A fairly common word. I noticed this well over a decade ago, and it's still an issue. Of course, I add it the first time I use it, so it's then checked correctly, but still.

The LibreOffice wiki has a list of the authors of the various dictionaries here: https://wiki.documentfoundation.org/Development/Dictionaries

**Reply by vkbo** [reply]
I've updated the documentation to include instructions on how to add dictionaries on Windows:
https://novelwriter.readthedocs.io/en/latest/int_customise.html#a-custom-dict

**Reply by jyhelle** [reply]
@Ellipsine, yes, I speak and write in French. 
To be honest, I don't use the spell checking feature while preparing the initial text version (I have made a living from proofreading books for several years so I can rather safely turn it off). 
Then, after exporting the text into an ODT file, I move to LibreOffice where I use the Grammalecte tool to check not only spelling but also verbs conjugation, typographical errors, words redundancy etc.  Have a look at the grammalecte.net page to see all it can do (alas, only for French texts)

To give you the broad picture, my workchain is as follows:
[plot ideas, research, storage of various bits and pieces] Obsidian
[world design, maps, etc if needed] WorldAnvil and/or Inkarnate
[text entry, general structure] novelWriter -> (odt) 
[spelling and grammar, page formatting, endnotes, typography, illustrations] LibreOffice and Grammalecte -> (odt) 
[conversion to epub format, sometimes Kindle ] Calibre -> (epub, azw3)
... at this time my "proofreading team" starts working for me ...
[epub format compliance] EpubCheck
[epub corrections] Sigil

**Reply by jyhelle** [reply]
@Ellipsine I checked [Firefox French dictionary](https://addons.mozilla.org/fr/firefox/addon/dictionnaire-fran%C3%A7ais1/) (not the language pack)  and since may 2020 it is provided by Olivier R., who is also the author of Grammalecte, so this dictionary should be OK since Firefox also uses Hunspell
(i don´t have Firefox at hand to check)

Edit: in fact there is a [Dicollecte](http://www.dicollecte.org/home.php?prj=fr) site to make proposals or to signal errors so the dictionnary properly evolves and you can download the various dictionaries from there

**Comment by Ellipsine** [comment]
@vkbo Thank you for adding this, and I know it has nothing to do with NovelWriter itself :) 

@jyhelle Thank you very much for this links! I don-t want to use Firefox in my worflow but I can try to extract it this time. I did not find Dicollecte, I hope it will work better ! :)

My workflow is more direct, I totaly skip the LibreOffice part as I dislike this software. Grammalecte itself adds typographical errors with inconsistences of spaces (sometimes it puts a thin one, sometimes no…), so I add them myself in Sublim Text while working my HTML code for Sigil. ST shows unicode of invisible characters, it is easier for me. Like you I use Calibre at the end to convert into anything.
